File: src/engine/actors/Camera.java
/*
* To change this license header, choose License Headers in Project Properties.
* To change this template file, choose Tools | Templates
* and open the template in the editor.
*/
package engine.actors;

import engine.pojo.Axis;
import engine.pojo.Constants;
import engine.core.Game;
import engine.actors.Actor;
import engine.core.Graphics;
import java.awt.Frame;
import java.awt.Graphics2D;

/**
 *
 * @author XiscoFerrer
 */
public class Camera extends Actor{
    
    public Axis axisConstrain = Axis.BOTH;
    private Actor followedObject = null;
    
    private float offsetLeft = 0;
    private float offsetUp = 0;
    private boolean follow = false;

    public Camera(Game gameRef) {
        super(gameRef);
    }
    
    @Override
    public void render(Graphics buffer, Frame frameToRender) {
    }
    
    @Override
    public void update(float deltaTime) {
        super.update(deltaTime);
        if(follow && followedObject != null){
            if(axisConstrain == Axis.BOTH || axisConstrain == Axis.HORIZONTAL){
                if ((followedObject.position.x - offsetLeft) - Constants.GAME_WIDTH / 2 > 0 && (followedObject.position.x + offsetLeft) + Constants.GAME_WIDTH / 2 < gameRef.worldBounds.x) {
                    position.x = followedObject.position.x - Constants.GAME_WIDTH / 2 - offsetLeft;
                }
            }
            if(axisConstrain == Axis.BOTH || axisConstrain == Axis.VERTICAL){
                if ((followedObject.position.y - offsetUp) - Constants.GAME_HEIGHT / 2 > 0 && (followedObject.position.y + offsetUp) + Constants.GAME_HEIGHT / 2 < gameRef.worldBounds.y) {
                    position.y = followedObject.position.y - Constants.GAME_HEIGHT / 2 - offsetUp;
                }
            }
            
        }
    }
    
    public void setFollowObject(Actor objectToFollow, float offsetLeft, float offsetUp){
        followedObject = objectToFollow;
        follow = true;
        this.offsetLeft = offsetLeft;
        this.offsetUp = offsetUp;
    }
    
    public void setFollowObject(Actor objectToFollow){
        followedObject = objectToFollow;
        follow = true;
        this.offsetLeft = 0;
        this.offsetUp = 0;
    }
    
}

File: ts/source/core/Camera.js
"use strict";
var XEngine;
(function (XEngine) {
    var AXIS;
    (function (AXIS) {
        AXIS["NONE"] = "none";
        AXIS["HORIZONTAL"] = "horizontal";
        AXIS["VERTICAL"] = "vertical";
        AXIS["BOTH"] = "both";
    })(AXIS = XEngine.AXIS || (XEngine.AXIS = {}));
    var Camera = (function () {
        function Camera(game) {
            this.game = game;
            this.position = new XEngine.Vector(0, 0);
            this.followedObject = null;
            this.axis = XEngine.AXIS.BOTH;
            this.pMatrix = mat4.create();
        }
        Camera.prototype.followObject = function (gameObject, offsetLeft, offsetUp) {
            this.follow = true;
            this.offsetLeft = offsetLeft || 0;
            this.offsetUp = offsetUp || 0;
            this.followedObject = gameObject;
        };
        Camera.prototype.update = function () {
            if (this.followedObject != null) {
                if (this.axis === XEngine.AXIS.BOTH || this.axis === XEngine.AXIS.HORIZONTAL) {
                    if ((this.followedObject.position.x - this.offsetLeft) - this.game.width / 2 > 0 &&
                        (this.followedObject.position.x + this.offsetLeft) + this.game.width / 2 < this.game.worldWidth) {
                        this.position.x = this.followedObject.position.x - this.game.width / 2 - this.offsetLeft;
                    }
                }
                if (this.axis === XEngine.AXIS.BOTH || this.axis === XEngine.AXIS.VERTICAL) {
                    if ((this.followedObject.position.y - this.offsetUp) - this.game.height / 2 > 0 &&
                        (this.followedObject.position.y + this.offsetUp) + this.game.height / 2 < this.game.worldHeight) {
                        this.position.y = this.followedObject.position.y - this.game.height / 2 - this.offsetUp;
                    }
                }
            }
            var right = this.game.width + this.position.x;
            var up = this.game.height + this.position.y;
            mat4.ortho(this.pMatrix, this.position.x, right, up, this.position.y, 0.1, 100);
        };
        return Camera;
    }());
    XEngine.Camera = Camera;
})(XEngine || (XEngine = {}));
